upgrade/v1beta1/upgrade.go

```diff
@@ -0,0 +1,214 @@
+	}
+
+	scheme := runtime.NewScheme()
+	if err := longhornV1beta1.SchemeBuilder.AddToScheme(scheme); err != nil {
+		return errors.Wrap(err, "unable to create scheme for v1beta1")
+	}
+
+	if err := upgradeVolumes(namespace, lhClientV1beta1, lhClient); err != nil {
+		return err
+	}
+	if err := upgradeEngineImages(namespace, lhClientV1beta1, lhClient); err != nil {
+		return err
+	}
+	if err := upgradeBackupTargets(namespace, lhClientV1beta1, lhClient); err != nil {
+		return err
+	}
+	if err := upgradeNodes(namespace, lhClientV1beta1, lhClient); err != nil {
+		return err
+	}
+
+	logrus.Infof("%v: completed", upgradeLogPrefix)
+	return nil
+}
+
+func copyConditions(fromConditions map[string]longhornV1beta1.Condition) ([]longhorn.Condition, error) {
+	toConditions := []longhorn.Condition{}
+	for _, fromCon := range fromConditions {
+		toCon := longhorn.Condition{}
+		if err := copier.Copy(&toCon, &fromCon); err != nil {
+			return nil, err
+		}
+		toConditions = append(toConditions, toCon)
+	}
+	return toConditions, nil
+}
+
+func CanUpgrade(config *restclient.Config, namespace string) (bool, error) {
+	lhClient, err := lhclientset.NewForConfig(config)
+	if err != nil {
+		return false, errors.Wrap(err, "unable to get clientset for v1beta1")
+	}
+
+	scheme := runtime.NewScheme()
+	if err := longhornV1beta1.SchemeBuilder.AddToScheme(scheme); err != nil {
+		return false, errors.Wrap(err, "unable to create scheme for v1beta1")
+	}
+
+	setting, err := lhClient.LonghornV1beta1().Settings(namespace).Get(context.TODO(), string(types.SettingNameDefaultEngineImage), metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// cannot find the setting,
+			logrus.Info("Longhorn CRD API v1beta1 not found")
+			return true, nil
+		}
+
+		return false, errors.Wrap(err, "unable to verify if version matches v1beta1")
+	}
+
+	return false, errors.Wrapf(err, "unable to upgrade from %v directly", setting.Value)
+}
+
+func upgradeVolumes(namespace string, lhClientV1beta1 *lhclientset.Clientset, lhClient *lhclientset.Clientset) error {
+	volumes, err := lhClientV1beta1.LonghornV1beta1().Volumes(namespace).List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return errors.Wrap(err, "unable to list volumes")
+	}
+	for _, old := range volumes.Items {
+		new := &longhorn.Volume{}
+
+		if err := copier.Copy(&new, &old); err != nil {
+			return err
+		}
+
+		conditions, err := copyConditions(old.Status.Conditions)
+		if err != nil {
+			return errors.Wrap(err, "upgrade: copy volume conditions")
+		}
+		new.Status.Conditions = conditions
+
+		new, err = lhClient.LonghornV1beta2().Volumes(namespace).UpdateStatus(context.TODO(), new, metav1.UpdateOptions{})
+		if err != nil {
+			if !apierrors.IsConflict(err) {
+				return errors.Wrapf(err, "failed to update for %v status %v", old.Kind, old.Name)
+			}
+			logrus.Warnf("%v: update status for %v %v v1beta1 result in conflict, skipping", upgradeLogPrefix, old.Kind, old.Name)
+			continue
```

Comment:
I knew these codes are derived from `upgrade/v1alpha1/upgrade.go`. But I don't think we can skip here if there is a conflict. Since later on we will update the API version setting to v1beta2 then there is no chance to redo this upgrade anymore.

Reply:
> > Leaving a not here.
> > For future upgrade refactoring, need to separate the two scenarios
> > (1) CRD version upgrade (2) CRD version not upgraded, but longhorn version upgrade
> > The logic of current upgrade mixes them up, so it is still messy now.
> 
> We always do API version upgrade then longhorn-version-based CR upgrade.
> 
> https://github.com/longhorn/longhorn-manager/blob/050cbf3d0a97c1c8d8681d611c9382bcb4b76daf/upgrade/upgrade.go#L113-L127
> 
> Maybe we can squash functions `doPodsUpgrade`, `doServicesUpgrade `, and `doDeploymentAndDaemonSetUpgrade` into `doCRUpgrade`. But I think the current flow in general is clear. Not sure which part could confuse you when you implement this PR.

There are upgrade CR in doAPIVersionUpgrade and doCRUpgrade, so it will be confusing. 
Yes, squashing functions is a good way to avoid confusion.

Reply:
It tries to UpgradeCRFromV1beta1ToV1beta2 first. The crdAPIVersionSetting is not updated if the first failed, so it will retry again next time.

```
	case types.CRDAPIVersionV1beta1:
		logrus.Infof("Upgrading from %v to %v", types.CRDAPIVersionV1beta1, types.CurrentCRDAPIVersion)
		if err := v1beta1.UpgradeCRFromV1beta1ToV1beta2(config, namespace, lhClient); err != nil {
			return err
		}
		crdAPIVersionSetting.Value = types.CRDAPIVersionV1beta2
		if _, err := lhClient.LonghornV1beta2().Settings(namespace).Update(context.TODO(), crdAPIVersionSetting, metav1.UpdateOptions{}); err != nil {
			return errors.Wrapf(err, "cannot finish CRD API upgrade by setting the CRDAPIVersionSetting to %v", types.CurrentCRDAPIVersion)
		}
```

Reply:
Sorry, this kind of volume update conflict won't fail `UpgradeCRFromV1beta1ToV1beta2`... The setting will be updated anyway...